Apply Better Practices in CMake

AlexBuccheri opened the pull request:
This MR implements the suggestions by @LecrisUT in issue #2

AlexBuccheri committed:
Remove non-standard dirs, and target MPI with the target provided by find_package

AlexBuccheri committed:
Assign flags via interface targets, to control scope

AlexBuccheri committed:
Clean up flag declarations (from testing)
Pass arguments to forunto exes, so each one runs ONLY its specific test

AlexBuccheri committed:
Change fortuno wrapper to be an OBJECT

AlexBuccheri committed:
Note some ctest args

AlexBuccheri committed:
Refactor code to more cleanly store centroid-grid mapping

AlexBuccheri commented on CMakeLists.txt:
```diff
@@ -2,39 +2,51 @@ cmake_minimum_required(VERSION 3.20)
 set(CMAKE_Fortran_MODULE_DIRECTORY ${PROJECT_BINARY_DIR}/modules)
 
 # Compiler flags for GCC only
+# Note, flags "should" be specified in yaml presets, and passed to cmake as arguments
```
I was too lazy to do that - in the next project

LecrisUT replied:
What "yaml preset"?

AlexBuccheri replied:
This is what you told me - flags should go in cmake presets, which are yaml format.

LecrisUT replied:
CMake presets are JSON, not yaml. How much I wish they had a yaml parser.

AlexBuccheri commented on CMakeLists.txt:
```diff
@@ -113,16 +118,16 @@ target_sources(lib_kmeans PUBLIC
 
 # Fortuno wrapper layer
-add_library(fortuno_wrapper "")
+add_library(fortuno_wrapper OBJECT "")
```
Could you elaborate on:

> This can be either an INTERFACE target (if it's used only once, otherwise you have duplicate source rules), or OBJECT library so that it does not actually create a library artifact.

It's linked to more than one binary, for example

LecrisUT replied:
`OBJECT` means that it owns the `.o` file, `INTERFACE` means that the consumer owns the `.o` file. If more than one target owns `.o` files with the same name, that is an error. `OBJECT` is basically the base class of `SHARED` and `STATIC`.

There are some nuances with some implied target properties. The usual one is [`POSITION_INDEPENDENT_CODE`](https://cmake.org/cmake/help/latest/prop_tgt/POSITION_INDEPENDENT_CODE.html)

AlexBuccheri commented on CMakeLists.txt:
```diff
@@ -132,24 +137,36 @@ target_sources(fortuno_wrapper PRIVATE
+# install(
+#    FILES ${CMAKE_CURRENT_BINARY_DIR}/kmeans.pc
+#    DESTINATION ${CMAKE_INSTALL_LIBDIR}/pkgconfig
+#    COMPONENT Fortuno_development
```
Oops haha

AlexBuccheri commented on CMakeLists.txt:
```diff
@@ -132,24 +137,36 @@ target_sources(fortuno_wrapper PRIVATE
+
+
+# Installation
+# --------------------------------
```
This is the next step, which I may also kindly ask you to grill

AlexBuccheri commented on CMakeLists.txt:
```diff
@@ -132,24 +137,36 @@ target_sources(fortuno_wrapper PRIVATE
-add_test(NAME test_linspace COMMAND test_grids)
-add_test(NAME test_linspace_to_grid2d COMMAND test_grids)
-add_test(NAME test_generate_gaussian_2d COMMAND test_grids)
+target_link_libraries(test_grids PRIVATE lib_kmeans fortuno_wrapper common_flags common_options) 
```
I knew there was a reason why each test binary was running all tests lol. Just had not got round to addressing it yet - thanks for the nudge

AlexBuccheri reviewed (commented):
**Other Points**

I didn't split another CMakeLists.txt into src because the project is so small.

`run.f90` is an app test... it's in a random place as I was previously having problems linking to it. This will move

LecrisUT reviewed (changes requested):
Will review later, just adding this to be marked as reviewer. Technically would not be authoritative reviewer for the context of github rules

LecrisUT commented on CMakeLists.txt:
```diff
@@ -2,39 +2,51 @@ cmake_minimum_required(VERSION 3.20)
 
 # Compiler flags for GCC only
+# Note, flags "should" be specified in yaml presets, and passed to cmake as arguments
+# Use c-preprocessor. Note, probably better to set target property `Fortran_PREPROCESS` instead
```
```suggestion
# Use c-preprocessor. Note, probably better to set target property `Fortran_PREPROCESS` instead
```

You don't need to do anything, it is automatic. Just rename the files `.f90` to `.F90`

LecrisUT commented on CMakeLists.txt:
```diff
@@ -2,39 +2,51 @@ cmake_minimum_required(VERSION 3.20)
+ #   -finit-real=nan                              # All real scalars are initialised to NaN
     )
-
+                    
```
```suggestion

```

Really?

LecrisUT commented on CMakeLists.txt:
```diff
@@ -2,39 +2,51 @@ cmake_minimum_required(VERSION 3.20)
+string(REPLACE " " ";" GCC_DEBUG "${GCC_DEBUG}")
+string(REPLACE " " ";" GCC_RELEASE "${GCC_RELEASE}")
+
+# Interface to control scope of compiler flags
+add_library(common_flags INTERFACE)
+# Control the scope of compiler options
+add_library(common_options INTERFACE)
```
Why separate?

LecrisUT commented on CMakeLists.txt:
```diff
@@ -2,39 +2,51 @@ cmake_minimum_required(VERSION 3.20)
+# list(JOIN GCC_BASE " " GCC_BASE)
+# set(CMAKE_Fortran_FLAGS "${CMAKE_Fortran_FLAGS} ${GCC_BASE}")  
+
+# When using an interface target to assign compiler flags (or options), one requires a semicolon-separated string
+# which is == strings appended to a list
+# Convert the space-separated string into a semicolon-separated string
+string(REPLACE " " ";" GCC_BASE "${GCC_BASE}")
+string(REPLACE " " ";" GCC_DEBUG "${GCC_DEBUG}")
+string(REPLACE " " ";" GCC_RELEASE "${GCC_RELEASE}")
```
What are you doing here? `set(var item_1 item_2 ....)` makes `var` a list already.

AlexBuccheri replied:
Took me a while to get the list in the correct form. I just tried removing these commands and the damn thing still works... no idea what was causing the problem

LecrisUT commented on CMakeLists.txt:
```diff
@@ -2,39 +2,51 @@ cmake_minimum_required(VERSION 3.20)
+add_library(common_options INTERFACE)
+
+# Build-specific flags
+if(CMAKE_BUILD_TYPE STREQUAL "Debug")
+   target_compile_options(common_flags INTERFACE ${GCC_BASE} ${GCC_DEBUG})
+elseif(CMAKE_BUILD_TYPE STREQUAL "Release")
+   target_compile_options(common_flags INTERFACE ${GCC_BASE} ${GCC_RELEASE})
+endif()
```
```suggestion
target_compile_options(common_flags INTERFACE ${GCC_BASE} 
    $<$<CONFIG:Debug>:${GCC_DEBUG}>
    $<$<CONFIG:Release>:${GCC_RELEASE}>
)
```

Explanation `$<CONFIG:Release>` converts to `0` or `1`, and `$<bool_value:some string>` expands `some string` if `bool_value == true`. It is always confusing how the `"` works around gen-ex, but trial-and-error.

You can go further with things like `$<Fortran_COMPILER_ID:GNU>`

AlexBuccheri replied:
Ooooo, clean af!

LecrisUT commented on CMakeLists.txt:
```diff
@@ -52,19 +64,11 @@ FetchContent_Declare(
-   message("-- Using MPI")
-
-   # Find MPI library and headers
    find_package(MPI REQUIRED)
```
```suggestion
   find_package(MPI REQUIRED Fortran)
```

LecrisUT commented on CMakeLists.txt:
```diff
@@ -82,10 +86,11 @@ if(MPI)
 else()
    set(MPI_Fortran_LIBRARIES "")
    message("-- Not using MPI")
+   # Create an empty target with the same name
+   add_library(MPI::MPI_Fortran INTERFACE IMPORTED)
```
You are shooting yourself in the foot here. You will only understand when you get to the packaging **and** you have some packaging tests.

AlexBuccheri replied:
I was getting compilation failure for serial code without this line. What is the correct thing to do?

LecrisUT replied:
What compilation error. Try to remove it and show it to me in the CI (you can link specific lines of the CI log)

LecrisUT commented on CMakeLists.txt:
```diff
@@ -96,13 +101,13 @@ set_target_properties(lib_kmeans
 
 # Linked libraries
-target_link_libraries(lib_kmeans PRIVATE OpenMP::OpenMP_Fortran ${MPI_Fortran_LIBRARIES} common_options)  # BLAS::BLAS 
+target_link_libraries(lib_kmeans PRIVATE OpenMP::OpenMP_Fortran MPI::MPI_Fortran common_flags common_options)
```
```suggestion
target_link_libraries(lib_kmeans PRIVATE
    OpenMP::OpenMP_Fortran
    $<$<BOOL:${MPI}>:MPI::MPI_Fortran>
    common_flags common_options
)
```

AlexBuccheri replied:
Ah, is this the correct thing to do? Never seen this before

LecrisUT commented on CMakeLists.txt:
```diff
@@ -96,13 +101,13 @@ set_target_properties(lib_kmeans
         PROPERTIES
         VERSION 0.0.1
         SOVERSION 0.0.1
```
```suggestion
        VERSION ${libkmeans_VERSION}
        SOVERSION ${libkmeans_VERSION_MAJOR}
```
Never set `SOVERSION == VERSION`. That's a middle-finger to packagers

AlexBuccheri replied:
I have never written cmake for packaging. I just set these out of convenience.

LecrisUT replied:
SOVERSION: Version of the cross-*compatible* libraries. I.e. when you link to a library it links to the specific SOVERSION used at build time. If you update a package, but the SOVERSION is identical, it will continue to work. But if you set it like that, you are indicating, no version is compatible with one-another.

LecrisUT reviewed (changes requested):
> I didn't split another CMakeLists.txt into src because the project is so small.

Shouldn't matter, best practices should be applied even to hello-world. I encourage you to do so and play around with it. There are nuances about CMake scopes that you can only understand by working and debugging these structures.

> run.f90 is an app test... it's in a random place as I was previously having problems linking to it. This will move

So basically `example/run.f90`. I would encourage to make it as a separate CMake project, that will help you figure out how to do packaging.

AlexBuccheri commented:
@LecrisUT Thanks for the second round of review - actually extremely useful. I will relocate the src definition, then push.
I'm not going to try the packaging on run.F90, I have another immediate use case.

AlexBuccheri committed:
Refactor to split into source-level and top-level CMakeLists.txt.
Replace if-statements and empty target with generator expressions.
Remove superfluous -cpp flag.
Remove superfluous flag list manipulation.
Consolidate common flag and option targets.
Use variables for static and SO versions.
